### ISSUE TITLE
How to load a Chrome Extension url

### ISSUE THREAD

**Issue by ORESoftware** [puppeteer/puppeteer#1796]
This is a support/feature/bug - I believe it's currently not possible to open a Chrome Extension URL with puppeteer...here is what I have:

```js
const puppeteer = require('puppeteer');   // version 1.0.0

(async () => {
  
  const args = puppeteer.defaultArgs().filter(arg => String(arg).toLowerCase() !== '--disable-extensions');

  const b = await puppeteer.launch({
    headless: false,
    devtools: true,
    ignoreDefaultArgs: true,
    args: args.concat(['--remote-debugging-port=9223'])
  });

  const browser = await puppeteer.connect({
    browserWSEndpoint:  b.wsEndpoint() ,   //`ws://${host}:${port}/devtools/browser/<id>`,
    ignoreHTTPSErrors: false
  });

  const page = await browser.newPage();
  await page.goto('chrome-extension://hfnplgaapcnjblpebnhibfcdfllkpaaf/dist/index.html');


})();
```

but I still can't launch the chrome extension, I get the same error:

> UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: net::ERR_BLOCKED_BY_CLIENT

there is no error if I do

`await page.goto('http://yahoo.com');`

so it has to do with attempting to load the extension, any ideas? I can load the extension url with regular Chrome, so the url is correct. Also, when I load Chromium, I don't see any extension icons. So my hypothesis is that I need to install/enable the extension in Chromium. For some reason I was thinking that it would just "copy" the extensions from my existing Chrome installation, but that was dumb thinking. Now I have to look up the API for installing a Chrome Extension...thanks for your help

Basically I am looking to launch a Chrome Extension page in a new tab, multiple times...the saga continues..

**Comment by ebidel** [comment]
There's info on how to launch chrome with an extension here: https://github.com/GoogleChrome/puppeteer/tree/master/examples#load-a-chrome-extension

However, Chrome doesn't allow you to navigate to some urls for security reasons. That includes `chrome-extension://`, `about:` and `chrome:` urls.

The best could be to make the extension open a new tab to its `index.html` when the extension loads for the first time.

**Comment by ORESoftware** [comment]
Thanks! I didn't quite follow your last sentence:

> The best could be to make the extension open a new tab to its index.html when the extension loads for the first time.

**Comment by ORESoftware** [comment]
Ok this worked for me! How exciting

```js
const puppeteer = require('puppeteer');

(async () => {

  const args = puppeteer.defaultArgs().filter(arg => String(arg).toLowerCase() !== '--disable-extensions');

  const browser = await puppeteer.launch({
    headless: false,
    devtools: true,
    ignoreDefaultArgs: true,
    args: args.concat([
      '--remote-debugging-port=9223',
      '--load-extension=/Users/alexamil/WebstormProjects/oresoftware/sumanjs/suman-chrome-extension'
      ]
    )
  });
  
  const c = await puppeteer.connect({
    browserWSEndpoint:  browser.wsEndpoint() ,   //`ws://${host}:${port}/devtools/browser/<id>`,
    ignoreHTTPSErrors: false
  });

  const page = await c.newPage();
  await page.goto('chrome-extension://hfnplgaapcnjblpebnhibfcdfllkpaaf/dist/index.html');

})();
```

one more question - is there a way to open DevTools at the bottom of the page instead of on the side?

**Comment by ORESoftware** [comment]
@ebidel now I follow your comment

> The best could be to make the extension open a new tab to its index.html when the extension loads for the first time.

The problem with that, is when developing, I cannot tell Chrome to open the new tab with DevTools open :(  Having DevTools open is the primary reason I want to use puppeteer to launch my Extension in a new tab.

**Comment by ebidel** [comment]
> However, Chrome doesn't allow you to navigate to some urls for security reasons. That includes chrome-extension://, about: and chrome: urls.

FWIW, I re-tested `page.goto('chrome-extension://...)` and chrome appears to be fine with this now. At least when using Puppeteer.

> I cannot tell Chrome to open the new tab with DevTools open 

When I launch chrome with`devtools: true`, _any_ tab that gets opened should auto open the devtools. That includes tabs opened by the extension using `chrome.tabs.create()`.

Out of curiosity, what's the reason you're using `puppeteer.connect` in that example?

**Comment by ORESoftware** [comment]
@ebidel thanks, the reason I am using puppeteer.connect is so that I can open a new tab in an existing window, instead of always opening a new window

```page.goto('chrome-extension://...)``` doesn't seem to work for me, but I could have been doing something wrong.

I found a solution that works for me for now. It involves what you recommended - the extension creates a new tab when the background page gets reloaded (when the extension is in development mode).

I can keep devtools open, by refreshing a tab instead of reloading the extension. I can refresh instead of reload if certain files change. But if the background script or content script changes, then I do have to reload the extension.

**Comment by ebidel** [comment]
> page.goto('chrome-extension://...) doesn't seem to work for me, but I could have been doing something wrong.

Yea, I feel like this shouldn't work, so maybe it's a regression in Chromium. I know it didn't before. Are you testing with puppeteer 1.0?

**Comment by ORESoftware** [comment]
yeah I installed v1.0.0, 2 days ago I think

**Comment by selaaa-online** [comment]
Can someone point me on the right direction on how to navigate to chrome extension popup using C# ?

```
await new BrowserFetcher().DownloadAsync(BrowserFetcher.DefaultRevision);
   string disableExtensionPath = $"--disable-extensions-except={args[0]}";
                    string loadExtensionPath = $"--load-extension={args[0]}";
                    string enableAutomation = "--enable-automation";
                    string noSandBox = "--no-sandbox";
                    string disableSetUidSandBox = "--disable-setuid-sandbox";
   string executablePath = "/usr/bin/chromium"; // executable path in ubuntu env
                    string[] arguments = new string[] { disableExtensionPath, loadExtensionPath, enableAutomation, disableSetUidSandBox, noSandBox, executablePath };

 var browser = await Puppeteer.LaunchAsync(new LaunchOptions { Headless = false, Args = arguments ,IgnoreHTTPSErrors=true, Devtools=true});

                    var c = await Puppeteer.ConnectAsync(new ConnectOptions{
                    BrowserWSEndpoint= browser.WebSocketEndpoint,
                     IgnoreHTTPSErrors= false
                    });

                    var page = await c.NewPageAsync();
                    Console.WriteLine("Enabling extension");
                    await page.GoToAsync("chrome-extension://gdocgbfmddcfnlnpmnghmjicjognhonm/popup.html");
                    await page.WaitForSelectorAsync("#start");
                    await page.FocusAsync("#start");
                    await page.ClickAsync("#start");
                    Console.WriteLine("Enabling extension successfull");
                    await page.WaitForTimeoutAsync(5000);

                    Console.WriteLine("Navigating to URL");
                    await page.GoToAsync("http://nebraskalegislature.gov/laws/browse-chapters.php?chapter=20");
                    Console.WriteLine("Finished navigatingto URL");
                    Console.WriteLine("Successfullly installed Proxy into Chromium");
                    await browser.CloseAsync();
                    Console.WriteLine("Closing browser!");
                    await Task.Delay(900000);
```
                    
This isnt working in when running on linux env. Runs fine in the local environment

**Comment by Triden** [comment]
> Может ли кто-нибудь указать мне правильное направление, как перейти к всплывающему окну расширения Chrome с помощью С#?
> 
> ```
> await new BrowserFetcher().DownloadAsync(BrowserFetcher.DefaultRevision);
>    string disableExtensionPath = $"--disable-extensions-except={args[0]}";
>                     string loadExtensionPath = $"--load-extension={args[0]}";
>                     string enableAutomation = "--enable-automation";
>                     string noSandBox = "--no-sandbox";
>                     string disableSetUidSandBox = "--disable-setuid-sandbox";
>    string executablePath = "/usr/bin/chromium"; // executable path in ubuntu env
>                     string[] arguments = new string[] { disableExtensionPath, loadExtensionPath, enableAutomation, disableSetUidSandBox, noSandBox, executablePath };
> 
>  var browser = await Puppeteer.LaunchAsync(new LaunchOptions { Headless = false, Args = arguments ,IgnoreHTTPSErrors=true, Devtools=true});
> 
>                     var c = await Puppeteer.ConnectAsync(new ConnectOptions{
>                     BrowserWSEndpoint= browser.WebSocketEndpoint,
>                      IgnoreHTTPSErrors= false
>                     });
> 
>                     var page = await c.NewPageAsync();
>                     Console.WriteLine("Enabling extension");
>                     await page.GoToAsync("chrome-extension://gdocgbfmddcfnlnpmnghmjicjognhonm/popup.html");
>                     await page.WaitForSelectorAsync("#start");
>                     await page.FocusAsync("#start");
>                     await page.ClickAsync("#start");
>                     Console.WriteLine("Enabling extension successfull");
>                     await page.WaitForTimeoutAsync(5000);
> 
>                     Console.WriteLine("Navigating to URL");
>                     await page.GoToAsync("http://nebraskalegislature.gov/laws/browse-chapters.php?chapter=20");
>                     Console.WriteLine("Finished navigatingto URL");
>                     Console.WriteLine("Successfullly installed Proxy into Chromium");
>                     await browser.CloseAsync();
>                     Console.WriteLine("Closing browser!");
>                     await Task.Delay(900000);
> ```
> 
> Это не работает при работе в Linux env. Отлично работает в локальной среде

Hello, did you manage to find a solution?

**Comment by SelakaKithmal** [comment]
> 

Yes was able find the solution

**Comment by SelakaKithmal** [comment]
`protected async Task<Browser> GetPuppeteerBrowser(bool headless, bool requireGeoUnblocking = false, string region = "USA")
        {
            Browser browser;

           
            string enableAutomation = "--enable-automation";
            string noSandBox = "--no-sandbox";
            string disableSetUidSandBox = "--disable-setuid-sandbox";
            string[] argumentsWithoutExtension = new string[] { "--disable-gpu", "--disable-dev-shm-usage", enableAutomation, disableSetUidSandBox, noSandBox };

            _logger.Information("Running with KProxy enabled mode in the Puppeteer Browser.");
            
            string baseDirectoryPath = Path.GetDirectoryName(this.GetType().Assembly.Location);
            _logger.Information("base directory path : {path}", baseDirectoryPath);
            string kproxyPath = Path.Combine(baseDirectoryPath, "artifacts", "kproxy");

            _logger.Information("kproxy path : {path}", kproxyPath);
            string disableExtensionPath = $"--disable-extensions-except={kproxyPath}";
            string loadExtensionPath = $"--load-extension={kproxyPath}";


            string[] argumentsWithExtension = new string[] { "--disable-gpu", "--disable-dev-shm-usage", disableExtensionPath, loadExtensionPath, enableAutomation, disableSetUidSandBox, noSandBox };

            browser = await Puppeteer.LaunchAsync(new LaunchOptions
            {
                Headless = false, // When loading extensions to browser it need to run with head.
                Args = argumentsWithExtension,
                IgnoreHTTPSErrors=true 
            });
            _logger.Information("running pupateer browser with version : {version}", await browser.GetVersionAsync());
            await UnblockBrowserForGeoRegionAsync(browser, region);

            return browser;
        }`

The above code worked for me

**Comment by ma-nk** [comment]
For anyone having similar issue: check if Chrome extension id is valid in automation browser, it is usually different in puppeteer profile from what is seen on your regular browser
